File: src/ram_push/scripts/rampush.py
#!/usr/bin/env python
import rospy
from std_msgs.msg import Float64
import os
import math

class ArmPushBox():
    def __init__(self):
        rospy.init_node('joint_controller', anonymous=True)
        deg = 0.017453293
        self.joint1pub = rospy.Publisher('/ram/joint1_position_controller/command', Float64, queue_size=10)
        self.joint2pub = rospy.Publisher('/ram/joint2_position_controller/command', Float64, queue_size=10)
        self.joint4pub = rospy.Publisher('/ram/joint4_position_controller/command', Float64, queue_size=10)
        self.joint6pub = rospy.Publisher('/ram/joint6_position_controller/command', Float64, queue_size=10)
        self.joint7pub = rospy.Publisher('/ram/joint7_position_controller/command', Float64, queue_size=10)
        self.rate = rospy.Rate(60)
        self.base_shoulder(0,3.62*deg, 0.46*deg)
        self.elbow_wrist(0,3.68*deg, 0.46*deg)
        rospy.sleep(0.5)
        self.push_object(0.05)
        rospy.sleep(1)
    
    def base_shoulder(self, current, vel1, vel2):
        self.j1 = current
        self.j2 = current
        for i in range(0, 26):
          self.j1 = current + vel1 * i
          self.j2 = current - vel2 * i
          self.joint1pub.publish(self.j1)
          self.joint2pub.publish(self.j2)
          print "joint1 position: %.2f" % (self.j1)
          print "joint2 position: %.2f" % (self.j2)
          rospy.sleep(0.05) 

    def elbow_wrist(self, current, vel1, vel2):
        self.j4 = current
        self.j6 = current 
        for i in range(0, 26):
          self.j4 = current - vel1 * i
          self.j6 = current + vel2 * i
          self.joint4pub.publish(self.j4)
          self.joint6pub.publish(self.j6)
          print "joint4 position: %.2f" % (self.j4)
          print "joint6 position: %.2f" % (self.j6)
          rospy.sleep(0.05)     
              
    def push_object(self, vel):
        start = 0
        for i in range(0, 11):
          self.j7 = start + vel * i
          self.joint7pub.publish(self.j7)
          print "joint7 position: %.2f" % (self.j7)
          rospy.sleep(0.05)  
                 
def main():
    obj = ArmPushBox()

if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass

File: src/ram_push/scripts/rampush.cpp
#include <ros/ros.h>
#include <std_msgs/Float64.h>
#include <math.h>
#include <vector>

// Defining the joint publishers
ros::Publisher joint1_pub;
ros::Publisher joint2_pub;
ros::Publisher joint4_pub;
ros::Publisher joint6_pub;
ros::Publisher joint7_pub;

const double PI = 3.14159265359;
void base_shoulder(double cur1, double cur2, double vel1, double vel2, bool ispositioning);
void elbow_wrist(double cur4, double cur6, double vel1, double vel2, bool ispositioning);
void push_object(double vel, bool ispositioning);
double degrees2radians(double angle_in_degrees);

int main(int argc, char **argv)
{
  // Initiate new ROS node named "joint_command_node"
  ros::init(argc, argv, "joint_command_node");
  ros::NodeHandle n;
  double cur1, cur2, cur4, cur6;
  double vel, vel1, vel2;
  bool ispositioning;

  // The joint Publishers
  joint1_pub = n.advertise<std_msgs::Float64>("/ram/joint1_position_controller/command", 10); 
  joint2_pub = n.advertise<std_msgs::Float64>("/ram/joint2_position_controller/command", 10);
  joint4_pub = n.advertise<std_msgs::Float64>("/ram/joint4_position_controller/command", 10);
  joint6_pub = n.advertise<std_msgs::Float64>("/ram/joint6_position_controller/command", 10);
  joint7_pub = n.advertise<std_msgs::Float64>("/ram/joint7_position_controller/command", 10);
  // Pushing Object
  base_shoulder(0.0,0.0, degrees2radians(92)/25, degrees2radians(-11.5)/25, true);
  elbow_wrist(0.0,0.0, degrees2radians(92)/25, degrees2radians(11.5)/25, true);
  ros::Duration(1).sleep();
  push_object(0.05, true);
  ros::Duration(2).sleep();
  // Rolling Back
  push_object(0.1, false);
  ros::Duration(0.5).sleep();
  elbow_wrist(degrees2radians(92),degrees2radians(11.5), degrees2radians(92)/10, degrees2radians(11.5)/10, false);
  ros::Duration(0.5).sleep();
  base_shoulder(degrees2radians(92), degrees2radians(-11.5), degrees2radians(92)/10, degrees2radians(-11.5)/10, false);  
}
double degrees2radians(double angle_in_degrees){
      return angle_in_degrees *PI /180.0;
}
void base_shoulder(double cur1, double cur2, double vel1, double vel2, bool ispositioning){
   std_msgs::Float64 j1;
   std_msgs::Float64 j2;
   ros::Rate loop_rate(60);	
   if (ispositioning){
    for (int i=0;i<25;i++){
     j1.data = cur1 + vel1 * i;
     joint1_pub.publish(j1); 
     j2.data = cur2 + vel2 * i;
     joint2_pub.publish(j2);
     ros::Duration(0.05).sleep();  
     ros::spinOnce(); 
     loop_rate.sleep();          
   }
  }
  else{
   for (int i=0;i<10;i++){
    j1.data = cur1 - vel1 * i;
    joint1_pub.publish(j1); 
    j2.data = cur2 - vel2 * i;
    joint2_pub.publish(j2);
    ros::Duration(0.05).sleep();  
   }
  }
}
void elbow_wrist(double cur4, double cur6, double vel1, double vel2, bool ispositioning){
   std_msgs::Float64 j4;
   std_msgs::Float64 j6;
   ros::Rate loop_rate(40);
   if (ispositioning){
    for (int i=0;i<25;i++){
     j4.data = cur4 - vel1 * i;
     joint4_pub.publish(j4); 
     j6.data = cur6 + vel2 * i;
     joint6_pub.publish(j6);
     ros::Duration(0.05).sleep();  
     ros::spinOnce(); 
     loop_rate.sleep();          
   }
  }
  else{
   for (int i=0;i<11;i++){
    j4.data = cur4 - vel1 * i;
    joint4_pub.publish(j4); 
    j6.data = cur6 - vel2 * i;
    joint6_pub.publish(j6);
    ros::Duration(0.05).sleep();          
   }
  }
}
void push_object(double vel, bool ispositioning){
  std_msgs::Float64 j7;
  double start = 0.0;
  ros::Rate loop_rate(60);
  if (ispositioning){
   for (int i=0;i<10;i++){
    j7.data = start + vel * i;
    joint7_pub.publish(j7); 
    ros::Duration(0.05).sleep();  
    ros::spinOnce(); 
    loop_rate.sleep();          
   }
  }
  else{
    for (int i=0;i<5;i++){
     j7.data = start - vel * i;
     joint7_pub.publish(j7); 
     ros::Duration(0.05).sleep();  
    }
  }
}
